src/Storages/MergeTree/DataPartStorageOnDisk.cpp

```diff
@@ -736,7 +737,13 @@ DataPartStoragePtr DataPartStorageOnDisk::clone(
         disk->removeRecursive(path_to_clone);
     }
     disk->createDirectories(to);
-    volume->getDisk()->copy(getRelativePath(), disk, to);
+    if (copy_content)
+    {
+        disk->createDirectories(path_to_clone);
+        volume->getDisk()->copyDirectoryContent(getRelativePath(), disk, path_to_clone);
```

Comment:
1) is it possible that src directory will be removed in between? (for example it was merged).
2) if you attach a partition containing a lot of parts, and during copying one of them 100% of disk space will be used - what will happen with tmp files, and with parts which were successfully attached?

Reply:
thank you for your suggestions. For q1, i think it is impossible, every time we call attach function, table will firstly stop merge and lock itself, then find all vaild parts and do the copy work. 
For q2, yes, we cannot garantee its atomicity.But we can reserve disk space to ensure part level copying work success. if all disks are exhausted, it will throw no enough space exception. Since currently there is no rollback logic inside, successed parts are still retained after copy work fails